### PR TITLE
Save more file information from initial call...

### PR DESCRIPTION
**Pull request by rocky**
In particular, resolve the initial source directory and store that as _Dbg_init_dir. Resolve the name of main bash program and add that to canonicalized filenames.

@jansorg I think this fixes the second problem you mentioned in #50 Please double check though.

### CODE REVIEW

**Comment by jansorg** [comment]
@rocky Unfortunately, this did not fix the problem.
I've attached my debug session, also showing file location and content.

**Comment by jansorg** [comment]
@rocky Thank you, this is fixing the problem!
Unfortunately, it seems to add a regression that break with a relative path doesn't work anymore, e.g. `break ./sources/lib.bash:2`.
`break`  with an absolute path is working well.

I've attached a screenshot of my session.

image

File: lib/hook.sh [hunk 1]
```diff
@@ -119,6 +119,13 @@ _Dbg_debug_trap_handler() {
 
     _Dbg_save_args "$@"
 
+    typeset _Dbg_full_filename
+    _Dbg_full_filename=$(_Dbg_is_file "${BASH_SOURCE[1]}")
```
Comment (jansorg) [inline lib/hook.sh]: nit: quoting
```suggestion
    _Dbg_full_filename="$(_Dbg_is_file "${BASH_SOURCE[1]}")"
```

File: bashdb.in [hunk 2]
```diff
@@ -102,6 +102,15 @@ if [[ -n $_Dbg_script_file ]] ; then
 fi
 typeset _Dbg_dollar_0="$0"
 
+# Resolve and save mapping for main script, and resolve
+# the starting directory.
+_Dbg_full_filename=$(_Dbg_is_file "$_Dbg_script_file")
```
Comment (jansorg) [inline bashdb.in]: nit: quoting
```suggestion
_Dbg_full_filename="$(_Dbg_is_file "$_Dbg_script_file")"
```

File: bashdb.in [hunk 3]
```diff
@@ -102,6 +102,15 @@ if [[ -n $_Dbg_script_file ]] ; then
...
+_Dbg_file2canonic["${_Dbg_script_file}"]="$_Dbg_full_filename"
+# Note: expand_filename is expanding a *directory* here, not a filename.
+# This might cause a problem in the future if _Dbg_expand_filename becomes
+# more specific about this aspect.
+_Dbg_init_cwd=$(_Dbg_expand_filename "${_Dbg_script_file%/*}")
```
Comment (jansorg) [inline bashdb.in]: nit: quoting
```suggestion
_Dbg_init_cwd="$(_Dbg_expand_filename "${_Dbg_script_file%/*}")"
```

**Comment by rocky** [comment]
> @rocky Thank you, this is fixing the problem! Unfortunately, it seems to add a regression that break with a relative path doesn't work anymore, e.g. `break ./sources/lib.bash:2`. `break` with an absolute path is working well.
> 
> I've attached a screenshot of my session.
> 
> image

Thanks for checking and the changes. I will try to look at and address this problem over the weekend.

**Comment by rocky** [comment]
> > @rocky Thank you, this is fixing the problem! Unfortunately, it seems to add a regression that break with a relative path doesn't work anymore, e.g. `break ./sources/lib.bash:2`. `break` with an absolute path is working well.
> > I've attached a screenshot of my session.
> > image
> 
> Thanks for checking and the changes. I will try to look at and address this problem over the weekend.

@jansorg When I went to look at this over the weekend the image was missing. Would you upload it again or give a sequence of instructions? 

<strike>When I debug I see that in relative path resolution favor the initial current working directory over the current working directory, and that can sometimes be wrong. But it is hard to know if this is exactly the problem. </strike> Nevermind. I see that this is only an image problem in my reply, not what you posted originally.

The key issue is that the "break" command is not finding the file when using a relative path. As you report, using an absolute path works fine.

**Comment by rocky** [comment]
@jansorg with 896a12f the example you give now works. However, this also disables the automatic file loading of files with relative file names not previously encountered. 

Instead, you must specify an explicit "load" here. The automatic loads of relative file names happened in the "list" and "break" commands. With this change, you will get an "internal error" message instead. 

Addressing that will take more work and effort. I am not sure though it is that important. What do you think?

**Comment by jansorg** [comment]
I'll be able to test your changes end of this week. Thank you for looking into this!

No automatic loading of relative paths seems okay to me, unless I misunderstood what how it works.
As far as I understand, instead of a "break ./a/b.sh" an additional "load ./a/b.sh" is needed first. I didn't even know that this was a thing and my IDE integration is always executing a "load" command first in such a case.

File: lib/fns.sh [hunk 4]
```diff
@@ -197,6 +197,9 @@ function _Dbg_linespec_setup {
   typeset -ri is_function=${word[1]}
   line_number=${word[0]}
   full_filename=$(_Dbg_is_file "$filename")
+  if [[ -z "$full_filename" ]] ; then
+      full_filename=$(_Dbg_resolve_expand_filename "$filename")
```
Comment (jansorg) [inline lib/fns.sh]: Is quoting needed here?

**Comment by rocky** [comment]
> I'll be able to test your changes end of this week. Thank you for looking into this!
> 
> No automatic loading of relative paths seems okay to me, unless I misunderstood what how it works. As far as I understand, instead of a "break ./a/b.sh" an additional "load ./a/b.sh" is needed first. I didn't even know that this was a thing and my IDE integration always executes a "load" command first in such a case.

In stepping, the debugger may encounter relative paths in BASH_SOURCE. It turns these into absolute paths and saves the mapping as soon as possible,  to avoid the problem of the current working directory changing inside the debugged program. So in that respect, the auto conversion always happens, and has to happen.

In addition, a user can issue debugger commands like "list" or "break"  fabricating any name the user wants.  It is here that I guess we hitting places where we are not converting the filename to its full name or storing canonical information about the file.

File: lib/fns.sh (hunk 4 above)
Comment (rocky) [reply lib/fns.sh]: I tested with quoting and it doesn't seem to hurt. Personally, I find the quotes within quotes confusing, and more generally the multiple substitution passes weird, but this is bash after all. 

So I've now made the change.

**Comment by jansorg** [comment]
I've tested your branch and as far as I can tell, it's working great! Thank you for fixing this problem!

**Pull request merged**